src/SourceBuild/content/eng/tools/tasks/Microsoft.DotNet.SourceBuild.Tasks.XPlat/UpdateNuGetConfigPackageSourcesMappings.cs

```diff
@@ -85,5 +169,149 @@ public override bool Execute()
+        {
+            // source-build-reference-package-cache is a dynamic source, populated by SBRP build,
+            // Discover all SBRP packages from checked in nuspec files.
+
+            if (string.IsNullOrEmpty(VmrRoot))
+            {
+                throw new InvalidDataException(string.Format(CultureInfo.CurrentCulture, "VmrRoot is not set - cannot determine SBRP packages."));
+            }
+
+            string sbrpRepoRoot = Path.Combine(VmrRoot, "src", "source-build-reference-packages");
+            if (!Directory.Exists(sbrpRepoRoot))
+            {
+                throw new InvalidDataException(string.Format(CultureInfo.CurrentCulture, "SBRP repo root does not exist in expected path: {0}", sbrpRepoRoot));
+            }
+
+            string[] nuspecFiles = Directory.GetFiles(sbrpRepoRoot, "*.nuspec", SearchOption.AllDirectories);
+            foreach (string nuspecFile in nuspecFiles)
+            {
+                try
+                {
+                    using Stream stream = File.OpenRead(nuspecFile);
+                    NupkgInfo info = GetNupkgInfo(stream);
+                    string id = info.Id.ToLower();
+                    string version = info.Version.ToLower();
+
+                    AddToHashtable(currentPackages, id, version);
+                    AddToHashtable(allSourcesPackages, SBRPCacheSourceName, id);
+                }
+                catch (Exception ex)
+                {
+                    throw new InvalidDataException(string.Format(CultureInfo.CurrentCulture, "Invalid nuspec file", nuspecFile), ex);
+                }
+            }
+        }
+
+        private void AddToHashtable(Hashtable hashtable, string key, string value)
+        {
+            if (hashtable.ContainsKey(key))
+            {
+                List<string> values = (List<string>)hashtable[key];
+                if (!values.Contains(value))
+                {
+                    values.Add(value);
+                }
+            }
+            else
+            {
+                hashtable.Add(key, new List<string> { value });
+            }
+        }
+
+        /// <summary>
+        /// Get nupkg info, id and version, from nupkg file.
+        /// </summary>
+        private NupkgInfo GetNupkgInfo(string path)
+        {
+            try
+            {
+                using Stream stream = File.OpenRead(path);
+                ZipArchive zipArchive = new(stream, ZipArchiveMode.Read);
+                foreach (ZipArchiveEntry entry in zipArchive.Entries)
+                {
+                    if (entry.Name.EndsWith(".nuspec"))
+                    {
+                        using Stream nuspecFileStream = entry.Open();
+                        return GetNupkgInfo(nuspecFileStream);
+                    }
+                }
+
+                throw new InvalidDataException(string.Format(CultureInfo.CurrentCulture, "Did not extract nuspec file from package: {0}", path));
```

Comment:
Throwing an exception in a try..catch is an anti-pattern and does not feed necessary here.  This could be moved outside the try..catch

Reply:
Fixed with https://github.com/dotnet/installer/pull/19114/commits/3f02bc95f3815f096425785f6fd3e2a351888a7e